cluster/test/client_transactions_test.rb

```diff
@@ -48,4 +48,25 @@ def test_cluster_client_does_not_support_transaction_by_multiple_keys
       assert_nil(redis.get("key#{i}"))
     end
   end
+
+  def test_cluster_client_does_support_transaction_with_optimistic_locking
+    redis.mset('{key}1', '1', '{key}2', '2')
+
+    another = Fiber.new do
+      cli = build_another_client
+      cli.mset('{key}1', '3', '{key}2', '4')
+      cli.close
+      Fiber.yield
+    end
+
+    redis.watch('{key}1', '{key}2') do |tx|
```

Comment:
I'm not sure this is the right interface that redis-cluster-client/redis-clustering should be exposing.

The normal redis-rb client interface looks like this:

```
redis.watch('{key}1', '{key}2') do
    redis.get('{key}1')
    redis.get('{key}2')
    # Make some decision about what to do based on the value of these keys
   redis.multi do |tx|
       tx.set('{key}1', 'new_v1')
       tx.set('{key}2', 'new_v2')
    end
end
```

The differences are...

* `watch` does not actually yield anything
* A transaction is not actually started unless `multi` is called

Would you like me to put up a different pair of PR's (one here, one in redis-cluster-client) to try and show how I think we could implement it in a compatible way?

Reply:
In my implementation, the block is finally passed to the redis-cluster-client gem. I think the redis-cluster-client gem should not be complex in excess due to the redis gem. I think the transaction feature is quite different between standalone and cluster. But of course, I think the other approaches are welcome.

Reply:
I think I can put most of the "hacking"/complexity in redis-clustering, but keep an interface in redis-cluster-client that actually makes sense in cluster mode.

Basically, redis-clustering is the "adapter" that takes the redis-cluster-client interface and makes it fit the redis-rb API pattern. That seems to make the most sense to me.

Let me try and get a couple of PR's up today :)

Reply:
@supercaracal what do you think about this approach? https://github.com/redis/redis-rb/pull/1256 and https://github.com/redis-rb/redis-cluster-client/pull/339

Basically, exposing the `OptimisticLocking` object from a dedicated `#watch` API on the redis-cluster-client side and then redis-rb makes that state implicit through the `@active_watcher` ivar.